### PR TITLE
refactor(rome_cli): file process

### PR DESCRIPTION
**Pull request by ematipico**
## Summary

I finally had a breakthrough! It took me a bit to arrive there, but I finally managed to breakdown the logic of the files and re-use it when needed.

There were two significant issues:
- the logic of processing a file lived in one single file, and that logic needed to consider different scenarios. This wasn't ideal because adding new logic causes more burden, and it's challenging to get it right without breaking the existing behaviour;
- it was difficult to move code around due to the lifetime constraints we have;

I created a new wrapper called `SharedTraversalOptions`. This tiny struct is used to make the compiler happy and tell it that "`ctx` is just borrowed, and it won't escape the boundaries of the lifetimes".

I created a new `WorkspaceFile`, another tiny wrapper meant to coordinate the writing of a new file. For example, when applying fixes to a file, we need to:
- write the new contents to disk;
- update the CST of the file we have in the workspace;

This must be done because when we pass to the formatting pass, we need to apply the format, we need to have the new content of the file at end, even for the diagnostics.

I will look into it later if it's possible to optimize this part. Unfortunately, we need to keep a `String` around (it was like this before too).

This is a **BREAKING CHANGE** because now **formatting diagnostics** are always shown when running `rome check`. If you noticed, many snapshots were updated, offering more diagnostics. **This is the intended feature of the `rome check` command**.

## Test Plan

I updated the existing tests and made sure that the emitted messages are correct.

### CODE REVIEW

File: crates/rome_cli/src/execute/process_file/workspace_file.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        self.input.as_str()
+    }
+
+    /// It updates the workspace file with `new_content`
+    pub(crate) fn update_file(&mut self, new_content: impl Into<String>) -> Result<(), Error> {
+        let new_content = new_content.into();
+        self.file
+            .set_content(new_content.as_bytes())
+            .with_file_path(self.path.display().to_string())?;
+        self.input = new_content.clone();
```
Comment (denbezrukov) [inline crates/rome_cli/src/execute/process_file/workspace_file.rs:71]: I'm wondering if there's duplication here, given that we maintain the file content in both the file and the input fields.

File: crates/rome_cli/src/execute/process_file/organize_imports.rs [hunk 2]
```diff
@@ -0,0 +1,45 @@
+use crate::execute::diagnostics::ResultExt;
+use crate::execute::process_file::workspace_file::WorkspaceFile;
+use crate::execute::process_file::{
+    DiffKind, FileResult, FileStatus, Message, SharedTraversalOptions,
+};
+use rome_diagnostics::category;
+use std::path::Path;
+
+/// Lints a single file and returns a [FileResult]
+pub(crate) fn _organize_imports<'ctx>(
```
Comment (denbezrukov) [inline crates/rome_cli/src/execute/process_file/organize_imports.rs:10]: Do we need `_` here?

Comment (ematipico) [reply crates/rome_cli/src/execute/process_file/organize_imports.rs:10]: yeah, other wise clippy says it's unused. I could remove it though :)

File: crates/rome_cli/src/execute/process_file/workspace_file.rs (hunk 1 above)
Comment (ematipico) [reply crates/rome_cli/src/execute/process_file/workspace_file.rs:71]: Yeah, that's the other optimisation I wanted to do. We have two files:
- one is a real one, which is stored in the file system
- the other is the one that lives in the workspace, where we store the CST

The two need to be in sync. Ideally, we could remove `self.input` from the struct, but that means that we would need to read the file from the file system every time, and I am not sure what's best: keep the content in memory or do more I/O operations.

We need to benchmark, which I can happily do once we land this PR.

Comment (denbezrukov) [reply crates/rome_cli/src/execute/process_file/workspace_file.rs:71]: Can we use `get_file_content`?  
https://github.com/rome/tools/blob/0affcbf3bc089c158e9644d918fc23d2984c0c03/crates/rome_service/src/workspace/server.rs#L325-L331

Comment (ematipico) [reply crates/rome_cli/src/execute/process_file/workspace_file.rs:71]: Completely missed that API! That's great, I will update the code

**Pull request merged**